File: Ruby/lib/add_two_numbers.rb
# frozen_string_literal: true

# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  dummy_head = ListNode.new
  tail = dummy_head
  carry = 0

  while l1 != nil || l2 != nil || carry != 0
    digit1 = l1 == nil ? 0 : l1.val
    digit2 = l2 == nil ? 0 : l2.val

    sum = digit1 + digit2 + carry
    digit = sum % 10
    carry = (sum / 10).floor

    new_node = ListNode.new(digit)
    tail.next = new_node
    tail = tail.next

    l1 = l1.next if !l1.nil?
    l2 = l2.next if !l2.nil?
  end

  return dummy_head.next
end

File: TypeScript/src/add_two_numbers.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null) {
  let dummyHead = new ListNode(0, null);
  let tail = dummyHead;
  let carry = 0;

  while (l1 !== null || l2 !== null || carry !== 0) {
    let digit1 = l1?.val ?? 0;
    let digit2 = l2?.val ?? 0;

    let sum = digit1 + digit2 + carry;
    let digit = sum % 10;
    carry = Math.floor(sum / 10);

    let newNode = new ListNode(digit, null);
    tail.next = newNode;
    tail = tail.next;

    l1 = l1?.next ?? null;
    l2 = l2?.next ?? null;
  }

  return dummyHead.next;
}

export { addTwoNumbers, ListNode };
